### ISSUE TITLE
CronjobTrigger retry logic appears faulty, doesn't allow for startup

### ISSUE THREAD

**Issue by delanni** [kubeless/cronjob-trigger#17]
**Is this a BUG REPORT or FEATURE REQUEST?**: BUG Report

**What happened**: When a `Function` and its driving `CronJobTrigger` are created together, chances are slim that the `CronJobTrigger` would find the corresponding function. There is supposed to be an exponentially increasing timeout in retrying but it seems that it's not working as intended. After a re-deploy to the `CronJobTrigger` resource, the function is found.

**What you expected to happen**: I expected the cronjob-trigger-controller to eventually find the function it should evoke, but it fails to start up.

**How to reproduce it (as minimally and precisely as possible)**: Set up Kubeless. Create a single `kubernetes` manifest file, or a `helm` chart that creates a function and a trigger together (real-life usecase is when setting up a cron-task, you'd co-locate the function and it's trigger in the same helm chart, for example). Try installing the chart/manifest. If it worked, try again. <50% chance I'm seeing that it's actually finding the function.

**Anything else we need to know?**:
I trimmed some irrelevant parts of the logs. I'm getting the impression that the exponential retry logic is not really working as intended, because the logs show that the 5x retries are burnt through within a second (the timestamps are the same).

```
time="2021-02-23T09:51:04Z" level=info msg="Running Kubeless cronjob trigger controller version: v1.0.3"
time="2021-02-23T09:51:04Z" level=info msg="Starting Cron Job Trigger con...
time="2021-02-23T09:51:04Z" level=info msg="Cronjob Trigger controller cac...
time="2021-02-23T09:51:04Z" level=info msg="Cron Job Trigger controller syn...
time="2021-02-23T10:04:29Z" level=info msg="Processing update to CronJo...
time="2021-02-23T10:04:29Z" level=error msg="Unable to find the function cleanup...
time="2021-02-23T10:04:29Z" level=error msg="Error processing kubeless/cleanup (will retry): ...
time="2021-02-23T10:04:29Z" level=info msg="Processing update to CronJob Trigg...
time="2021-02-23T10:04:29Z" level=error msg="Unable to find the function clea..
time="2021-02-23T10:04:29Z" level=error msg="Error processing kubeless/cleanup (will retry): ...
time="2021-02-23T10:04:29Z" level=info msg="Processing update to CronJo...
time="2021-02-23T10:04:29Z" level=error msg="Unable to find the function cle...
time="2021-02-23T10:04:29Z" level=error msg="Error processing kubeless/cleanup (will retry): ...
...
time="2021-02-23T10:04:29Z" level=error msg="Error processing kubeless/cleanup (will retry): ...
time="2021-02-23T10:04:29Z" level=info msg="Processing update to CronJob Tr...
time="2021-02-23T10:04:29Z" level=error msg="Unable to find the function cleanup...
time="2021-02-23T10:04:29Z" level=error msg="Error processing kubeless/cleanup (will retry): ...
time="2021-02-23T10:04:29Z" level=info msg="Processing update to CronJo...
time="2021-02-23T10:04:29Z" level=error msg="Unable to find the function cleanup...
time="2021-02-23T10:04:29Z" level=error msg="Error processing kubeless/cleanup (giving up): ...
E0223 10:04:29.445826       1 cronjob_trigger_controller.go:186] functions.kubeless.io "cleanup" not found
```

**Environment**:
- Kubernetes version (use `kubectl version`): v1.18.10
- Kubeless version (use `kubeless version`): v1.0.8
- Cloud provider or physical cluster: AWS Kubernetes

**Comment by andresmgot** [comment]
Hi @delanni, 

I agree, that's weird, we should probably check the retry logic there. To unblock you, maybe you can try to use helm hooks to install both resources at different stages? You can try to install the CronJobTrigger in the post-install hook for example.

**Comment by delanni** [comment]
Thanks for the suggestions @andresmgot! I'll check out what I can do with hooks for now. Our plan with Kubeless is to simplify our architecture a bit, but if we have to hack around, it wouldn't result in altogether optimization :) 

Having looked into the issue, it's probably not a proper bug, simply a bit over-optimistic config:
https://github.com/kubeless/cronjob-trigger/blob/93203511328d6cb770e04067169a19155fe4cbc2/vendor/k8s.io/client-go/util/workqueue/default_rate_limiters.go#L41

Starting out with 5ms, if we're upping to the 4th exponent, it's still only 2^4*5 = 80ms, with the sum total of <160ms. 

If you agree with an adjustment in this area, I'd either suggest retrying ~10x (so the change is only local to the specific creation retries) or adjust the base timeout to 150ms (which would mean a total wait of 5s, which is still OK in human time).

What do you think? Maybe this could be a config param?

**Comment by delanni** [comment]
Yes, unfortunately, with helm's hooks, I'd need to have some executors prepared with `kubeless` and potentially `kubectl` prepared on an image so I can fire off the setup request afterwards, so it's not an option for the simplification :)

**Comment by andresmgot** [comment]
right, we could either increase that base timeout or just increase the number of retries:

https://github.com/kubeless/cronjob-trigger/blob/master/pkg/controller/cronjob_trigger_controller.go#L43

If we do 10 retries rather than 5, the wait step goes up until ~5s, which should be more than enough. Would you mind sending a PR? The CI can build an image that you can use afterwards to check if that will fix your issue.

**Comment by andresmgot** [comment]
> Yes, unfortunately, with helm's hooks, I'd need to have some executors prepared with `kubeless` and potentially `kubectl` prepared on an image so I can fire off the setup request afterwards, so it's not an option for the simplification :)

mmh, that should not be needed. You would create the same manifests, just adding the hook annotation for the trigger to be created after the rest is installed: 

https://helm.sh/docs/topics/charts_hooks/

But maybe I am unaware of some requirements of your setup.

**Comment by delanni** [comment]
You're right. I thought I should try to create the resource with the `kubeless` cli in a post-install hook (as if post-install hooks were scripts, like in most other systems I know). But in fact, the hooks here are resources, so I can fit the triggers there. I'll give that a shot, thanks for recommending! :)

**Comment by delanni** [comment]
I missed your first suggestion, I'll create a PR soon!

**Comment by andresmgot** [comment]
@delanni can you try this image: `docker.io/kubeless/cronjob-trigger-controller:build-da98b0a0-20bf-42da-8b4c-3bc1af6631a4`?

To double check it fixes your issue.
